util/pingpong.c

```diff
@@ -0,0 +1,2111 @@
+#define FT_ACTIVATE_DEBUG _activate_debug
+
+#define FT_DEBUG(fmt, ...)						\
+	do {								\
+		if (_activate_debug) {					\
+			FT_LOG_("debug", fmt, ##__VA_ARGS__);		\
+		}							\
+	} while (0)
+
+#define FT_EQ_ERR(eq, entry, buf, len) \
+	FT_ERR("eq_readerr: %s", fi_eq_strerror(eq, entry.prov_errno, \
+				entry.err_data, buf, len))
+
+#define FT_CQ_ERR(cq, entry, buf, len) \
+	FT_ERR("cq_readerr: %s", fi_cq_strerror(cq, entry.prov_errno, \
+				entry.err_data, buf, len))
+
+#define FT_CLOSE_FID(fd)						\
+	do {								\
+		int ret;						\
+		if ((fd)) {						\
+			ret = fi_close(&(fd)->fid);			\
+			if (ret)					\
+				FT_ERR("fi_close (%d) fid %d",		\
+					ret, (int) (fd)->fid.fclass);	\
+			fd = NULL;					\
+		}							\
+	} while (0)
+
+#define FT_CLOSEV_FID(fd, cnt)			\
+	do {					\
+		int i;				\
+		if (!(fd))			\
+			break;			\
+		for (i = 0; i < (cnt); i++) {	\
+			FT_CLOSE_FID((fd)[i]);	\
+		}				\
+	} while (0)
+
+
+#define FT_PROCESS_QUEUE_ERR(readerr, rd, queue, fn, str)	\
+	do {							\
+		if (rd == -FI_EAVAIL) {				\
+			readerr(queue, fn " " str);		\
+		} else {					\
+			FT_PRINTERR(fn, rd);			\
+		}						\
+	} while (0)
+
+#define FT_PROCESS_EQ_ERR(rd, eq, fn, str) \
+	FT_PROCESS_QUEUE_ERR(eq_readerr, rd, eq, fn, str)
+
+#define FT_PRINT_OPTS_USAGE(opt, desc) fprintf(stderr, " %-20s %s\n", opt, desc)
+
+#define MIN(a,b) (((a)<(b))?(a):(b))
+#define MAX(a,b) (((a)>(b))?(a):(b))
+#define ARRAY_SIZE(A) (sizeof(A)/sizeof(*A))
+
+#define TEST_ENUM_SET_N_RETURN(str, enum_val, type, data)	\
+	TEST_SET_N_RETURN(str, #enum_val, enum_val, type, data)
+
+#define TEST_SET_N_RETURN(str, val_str, val, type, data)	\
+	do {							\
+		if (!strncmp(str, val_str, strlen(val_str))) {	\
+			*(type *)(data) = val;			\
+			return 0;				\
+		}						\
+	} while (0)
+
+
+#define CTRL_BUF_LEN 64
+
+struct ct_pingpong {
+	struct fi_info *fi_pep, *fi, *hints;
+	struct fid_fabric *fabric;
+	struct fid_domain *domain;
+	struct fid_pep *pep;
+	struct fid_ep *ep;
+	struct fid_cq *txcq, *rxcq;
+	struct fid_mr *mr;
+	struct fid_av *av;
+	struct fid_eq *eq;
+
+	struct fid_mr no_mr;
+	struct fi_context tx_ctx, rx_ctx;
+	struct fi_context *ctx_arr;
+	uint64_t remote_cq_data;
+
+	uint64_t tx_seq, rx_seq, tx_cq_cntr, rx_cq_cntr;
+	//int ft_skip_mr;
+	int ft_parent_proc;
+	pid_t ft_child_pid;
+
+	fi_addr_t remote_fi_addr;
+	void *buf, *tx_buf, *rx_buf;
+	size_t buf_size, tx_size, rx_size;
+	int rx_fd, tx_fd;
+	int data_default_port;
+	char data_port[8];
+
+	char test_name[50];
+	int timeout;
+	struct timespec start, end;
+
+	struct fi_av_attr av_attr;
+	struct fi_eq_attr eq_attr;
+	struct fi_cq_attr cq_attr;
+	struct ft_opts opts;
+
+	long cnt_ack_msg;
+
+	int ctrl_port;
+	int ctrl_listenfd;
+	int ctrl_connfd;
+	struct sockaddr_in ctrl_addr;
+	char ctrl_buf[CTRL_BUF_LEN + 1];
+};
+
+#define FT_ENABLE_ALL		(~0)
+#define FT_DEFAULT_SIZE		(1 << 0)
+
+struct test_size_param {
+	int size;
+	int enable_flags;
+};
+
+struct test_size_param test_size[] = {
```

Comment:
> At first sight, I had the same reaction, but it is actually simple : it is a selector of message size. If you test the default sizes with usnic, it won't test them all, because providers do not accept all sizes.

I didn't say I didn't know what it did, I know exactly what it does. I just don't like how it does it. I'm not sure anything I come up with will be any better. First thought is something like this: 

```c
int *generate_test_data(struct ft_opts *options, size_t provider_maximum)
{
        int defaults[6] = {64, 256, 1024, 4096, 655616, 1048576};
        int power_of_two;
        int half_up;
        int *sizes;

        sizes = calloc(64, sizeof(*sizes));
        if (!sizes)
                return NULL;

        if (options->options & FT_OPT_SIZE) {
                if (options->transfer_size > provider_maximum)
                        return sizes;

                sizes[0] = options->transfer_size;
        } else if (options->sizes_enabled != FT_ENABLE_ALL) {
                for (int i = 0; i < sizeof(defaults); i++) {
                        if (defaults[i] > provider_maximum)
                                break;

                        sizes[i] = defaults[i];
                }
        } else {
                for (int i = 0; i < 22; i++) {
                        power_of_two = (1 << (i + 1));
                        half_up = power_of_two + (power_of_two / 2);

                        if (power_of_two > provider_maximum)
                                break;

                        sizes[i * 2] = power_of_two;

                        if (half_up > provider_maximum)
                                break;

                        sizes[(i * 2) + 1] = half_up;
                }
        }

        return sizes;
}
```

It does the same thing, but is it more readable?